### DISCUSSION TITLE
A gleam shell

### DISCUSSION THREAD

**Discussion by vincent-lg** [gleam-lang/gleam#1305] [Ideas & suggestions]
At this time, Gleam offers a `gleam shell` command which opens an Erlang shell.  Beyond the obvious (testing requires to know Erlang), I would say a shell is great even when one doesn't have a project yet: learning a programming language with a proper REPL can make a huge difference.  The ability offered by languages like Elixir to not only open a shell, but also to recompile modules on the fly makes testing syntax much easier.

My apologies if this feature has already been discussed, I couldn't find a link.

The floor is yours!

**Comment by lpil** [comment]
I think this would be a great thing to have. Let's discuss here how we might implement this and if someone wants to take ownership of this feature I would be very keen to give them any guidancing within the compiler.

**Comment by ssaunderss** [comment]
What would be the functionality wish list for this?

I have a few:

- Being able to recompile the project and individual files on the fly - @lpil I'm not too familiar with Gleam's compiler, does it recompile everything or just changed files?
- Friendlier syntax (Possibly drop Erlang shell syntax for something else - I'm not sure if Elixir's syntax would make sense in the shell (`MyApp.SomeFolder.SomeFile`) since Gleam already uses forward slash for import syntax)
- Ability to alias modules

I'm also willing to take a stab at this, but just a heads up it might take a while because I'm going to be learning Rust along the way 😀

**Reply by lpil** [reply]
I think just being able to evaluate Gleam expressions without any recompilation or extra features would be a great first step.

**Reply by suraj-mandal** [reply]
I have to type erlang code in the gleam shell, and I think basically just for interacting with the shell, I have to learn whole new language. Is it possible to somehow type gleam code directly in the shell and get the output?

**Reply by lpil** [reply]
Hey @suraj-mandal ! Currently we don't have a Gleam shell so you'll have to use the Erlang or JavaScript ones.

**Comment by bobjoe12131** [comment]
I remember a rust program (i dont remember the name) that was a rust "REPL" by taking in lines and putting them in the correct spots in a temp file, and then compiling it. Maybe a tool like that could work in the meantime.

**Reply by lpil** [reply]
This would result in the same side effects being emitted each time a new line is added. Given a REPL is commonly used for side effects this would result in the wrong thing being done and could potentially cause very serious problems for people using it expecting it to work like a REPL.
